cmd/tsbs_generate_queries/databases/clickhouse/devops_test.go

```diff
@@ -84,3 +88,443 @@ func TestDevopsGetSelectClausesAggMetrics(t *testing.T) {
 		}
 	}
 }
+
+func TestMaxAllCPU(t *testing.T) {
+	cases := []testCase{
+		{
+			desc:    "negative hosts",
+			input:   -1,
+			fail:    true,
+			failMsg: "number of hosts cannot be < 1; got -1",
+		},
+		{
+			desc:    "zero hosts",
+			input:   0,
+			fail:    true,
+			failMsg: "number of hosts cannot be < 1; got 0",
+		},
+		{
+			desc:               "1 host",
+			input:              1,
+			expectedHumanLabel: "ClickHouse max of all CPU metrics, random    1 hosts, random 8h0m0s by 1h",
+			expectedHumanDesc: "ClickHouse max of all CPU metrics, random    1 hosts, " +
+				"random 8h0m0s by 1h: 1970-01-01T00:47:30Z",
+			expectedQuery: `
+        SELECT
+            toStartOfHour(created_at) AS hour,
+            max(usage_user) AS max_usage_user, max(usage_system) AS max_usage_system, max(usage_idle) AS max_usage_idle, max(usage_nice) AS max_usage_nice, max(usage_iowait) AS max_usage_iowait, max(usage_irq) AS max_usage_irq, max(usage_softirq) AS max_usage_softirq, max(usage_steal) AS max_usage_steal, max(usage_guest) AS max_usage_guest, max(usage_guest_nice) AS max_usage_guest_nice
+        FROM cpu
+        WHERE (hostname = 'host_5') AND (created_at >= '1970-01-01 00:47:30') AND (created_at < '1970-01-01 08:47:30')
+        GROUP BY hour
+        ORDER BY hour
+        `,
+		},
+		{
+			desc:               "5 hosts",
+			input:              5,
+			expectedHumanLabel: "ClickHouse max of all CPU metrics, random    5 hosts, random 8h0m0s by 1h",
+			expectedHumanDesc:  "ClickHouse max of all CPU metrics, random    5 hosts, random 8h0m0s by 1h: 1970-01-01T00:17:45Z",
+			expectedQuery: `
+        SELECT
+            toStartOfHour(created_at) AS hour,
+            max(usage_user) AS max_usage_user, max(usage_system) AS max_usage_system, max(usage_idle) AS max_usage_idle, max(usage_nice) AS max_usage_nice, max(usage_iowait) AS max_usage_iowait, max(usage_irq) AS max_usage_irq, max(usage_softirq) AS max_usage_softirq, max(usage_steal) AS max_usage_steal, max(usage_guest) AS max_usage_guest, max(usage_guest_nice) AS max_usage_guest_nice
+        FROM cpu
+        WHERE (hostname = 'host_9' OR hostname = 'host_5' OR hostname = 'host_1' OR hostname = 'host_7' OR hostname = 'host_2') AND (created_at >= '1970-01-01 00:17:45') AND (created_at < '1970-01-01 08:17:45')
+        GROUP BY hour
+        ORDER BY hour
+        `,
+		},
+		{
+			desc:    "more hosts then cardinality (11)",
+			input:   11,
+			fail:    true,
+			failMsg: "number of hosts (11) larger than total hosts. See --scale (10)",
+		},
+	}
+
+	testFunc := func(d *Devops, c testCase) query.Query {
+		q := d.GenerateEmptyQuery()
+		d.MaxAllCPU(q, c.input)
+		return q
+	}
+
+	start := time.Unix(0, 0)
+	end := start.Add(devops.MaxAllDuration).Add(time.Hour)
+
+	runTestCases(t, testFunc, start, end, cases)
+}
+
+func diffStrings(first, second string) int {
+	firstBytes := []byte(first)
+	secondBytes := []byte(second)
+
+	for i, byte := range firstBytes {
+		if byte != secondBytes[i] {
```

Comment:
small nit: `b` might be nicer so that its not shadowing the type `byte`

or you could just do `for i := range firstBytes` since you only care about the index

Reply:
That was supposed to be a short debug function, I'll remove it completely.